driver-core/src/main/java/com/datastax/driver/core/ControlConnection.java

```diff
@@ -289,7 +291,13 @@ public void refreshSchema(String keyspace, String table) throws InterruptedExcep
             // At startup, when we add the initial nodes, this will be null, which is ok
             if (c == null)
                 return;
-            refreshSchema(c, keyspace, table, cluster, false);
+            refreshSchema(c, keyspace, table, cluster);
+            // If the table is null, we either rebuild all from scratch or have an updated keyspace. In both cases, rebuild the token map
+            // since some replication on some keyspace may have changed
+            if (table == null) {
+                cluster.submitNodeListRefresh();
+                cluster.nodeListRefreshRequestDebouncer.deliverEvents().get();
```

Comment:
Was able to reproduce this case by creating a keyspace when NON_BLOCKING_EXECUTOR_SIZE is 2.  The future never completes.  Added test case and analysis to this [gist](https://gist.github.com/tolbertam/8c024d5e6fbae1401bea).